### PR TITLE
8266404: Fatal error report generated with -XX:+CrashOnOutOfMemoryError should not contain suggestion to submit a bug report

### PR DESCRIPTION
**Pull request by dholmes-ora**
A simple fix to add a new VMErrorType (OOM_JAVA_HEAP_FATAL) to indicate we've requested a Java OOM error to be fatal. A new overload of report_fatal is provided to allow this to be passed through.

VMError has an existing notion of "should_report_bug" but that encompasses more than just printing the bug submission URL, so I added a new specific check for that.

Checked hs_err files before and after with the fix, from the CrashOnOutOfMemoryError test.

Tested tiers 1-3 for good measure.

Thanks,
David

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266404](https://bugs.openjdk.java.net/browse/JDK-8266404): Fatal error report generated with -XX:+CrashOnOutOfMemoryError should not contain suggestion to submit a bug report

### Reviewers
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to 147b5cf10be7c3bd2a5b8d6fd4185ba596fef4e8
 * [Kevin Walls](https://openjdk.java.net/census#kevinw) (@kevinjwalls - Committer) ⚠️ Review applies to 147b5cf10be7c3bd2a5b8d6fd4185ba596fef4e8
 * [Gerard Ziemski](https://openjdk.java.net/census#gziemski) (@gerard-ziemski - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4006/head:pull/4006` \
`$ git checkout pull/4006`

Update a local copy of the PR: \
`$ git checkout pull/4006` \
`$ git pull https://git.openjdk.java.net/jdk pull/4006/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4006`

View PR using the GUI difftool: \
`$ git pr show -t 4006`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4006.diff">https://git.openjdk.java.net/jdk/pull/4006.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/utilities/debug.cpp [hunk 1]
```diff
@@ -288,11 +288,15 @@ void report_vm_status_error(const char* file, int line, const char* error_msg,
   report_vm_error(file, line, error_msg, "error %s(%d), %s", os::errno_name(status), status, detail);
 }
 
-void report_fatal(const char* file, int line, const char* detail_fmt, ...)
-{
+// Declaration for ATTRIBUTE_PRINTF application.
+void report_fatal_impl(VMErrorType error_type, const char* file, int line,
+                       const char* detail_fmt, va_list detail_args)
+  ATTRIBUTE_PRINTF(4,0);
```
Comment (tstuefe) [inline src/hotspot/share/utilities/debug.cpp:294]: Can prototype and implementation be merged while retaining the attrib marker?

File: src/hotspot/share/utilities/debug.cpp [hunk 2]
```diff
@@ -288,11 +288,15 @@ void report_vm_status_error(const char* file, int line, const char* error_msg,
...
+
+// Definition
+void report_fatal_impl(VMErrorType error_type, const char* file, int line,
+                       const char* detail_fmt, va_list detail_args) {
```
Comment (tstuefe) [inline src/hotspot/share/utilities/debug.cpp:298]: Can this function be static? Any reason to expose it?

File: src/hotspot/share/utilities/debug.hpp [hunk 3]
```diff
@@ -150,7 +150,8 @@ enum VMErrorType {
   INTERNAL_ERROR   = 0xe0000000,
   OOM_MALLOC_ERROR = 0xe0000001,
   OOM_MMAP_ERROR   = 0xe0000002,
-  OOM_MPROTECT_ERROR = 0xe0000003
+  OOM_MPROTECT_ERROR = 0xe0000003,
+  OOM_JAVA_HEAP_FATAL = 0xe000004
```
Comment (tstuefe) [inline src/hotspot/share/utilities/debug.hpp:154]: Add one more zero? To keep it in line with the other errors?

File: src/hotspot/share/utilities/debug.hpp [hunk 4]
```diff
@@ -164,6 +165,7 @@ void report_vm_error(const char* file, int line, const char* error_msg,
 void report_vm_status_error(const char* file, int line, const char* error_msg,
                             int status, const char* detail);
 void report_fatal(const char* file, int line, const char* detail_fmt, ...) ATTRIBUTE_PRINTF(3, 4);
+void report_fatal(VMErrorType error_type, const char* file, int line, const char* detail_fmt, ...) ATTRIBUTE_PRINTF(4, 5);
```
Comment (tstuefe) [inline src/hotspot/share/utilities/debug.hpp:168]: See above, can we not expose this?

**Review by tstuefe — CHANGES_REQUESTED** [review]
Hi David,

mostly fine, minor nits.

Side note, I really would like to clean up this coding a bit. I have attempted to so so multiple times, but the patches ballooned. I may try again, maybe in tiny baby steps.

Cheers, Thomas

**Review by tstuefe — APPROVED** [review]
All good.

**Comment by kevinjwalls** [comment]
Hi David, looks good.

I get it now I've found that should_report_bug() really means "is vm internal error".  Literally, "is this NOT a native memory allocation failure".

If you had time to rename should_report_bug() "is_internal_error" or similar, I think that would make vmError.cpp significantly more readable.  (It's not a problem in the function you're adding, it's the old should_report_bug that seems misnamed, and more so now it has a similar related function.)

**Review by kevinjwalls — APPROVED** [review]

**Comment by tstuefe** [comment]
Hi Kevin,

> If you had time to rename should_report_bug() "is_internal_error" or similar, I think that would make vmError.cpp significantly more readable. (It's not a problem in the function you're adding, it's the old should_report_bug that seems misnamed, and more so now it has a similar related function.)

My preference would be to keep the patches as concise as possible, because that makes backporting patches easier. vmError is overdue for some cleanups, but I'd prefer those concentrated in a cleanup RFE, which would give us just one commit to packport, instead of spreading cleanups over unrelated patches.

Just my 5c, David may see this differently.

Cheers, Thomas

File: src/hotspot/share/utilities/debug.cpp [hunk 5]
```diff
@@ -302,7 +306,22 @@ void report_fatal(const char* file, int line, const char* detail_fmt, ...)
...
+                          Thread::current_or_null(), NULL, NULL, context,
+                          file, line, 0);
+}
+
+void report_fatal(const char* file, int line, const char* detail_fmt, ...) {
+  va_list detail_args;
+  va_start(detail_args, detail_fmt);
+  report_fatal_impl(INTERNAL_ERROR, file, line, detail_fmt, detail_args);
+  va_end(detail_args);
+}
```
Comment (gerard-ziemski) [inline src/hotspot/share/utilities/debug.cpp:319]: Do we really need this API? Can't we just update:

```
#define fatal(...)                                                                \
do {                                                                              \
  TOUCH_ASSERT_POISON;                                                            \
  report_fatal(INTERNAL_ERROR, __FILE__, __LINE__, __VA_ARGS__);                                  \
  BREAKPOINT;                                                                     \
} while (0)

```

**Review by gerard-ziemski — CHANGES_REQUESTED** [review]

File: src/hotspot/share/utilities/debug.cpp (hunk 5 above)
Comment (dholmes-ora) [reply src/hotspot/share/utilities/debug.cpp:319]: As the fatal() macro seems to be the only user of the existing `report_fatal` then yes I think we can simplify this as you suggest. I had thought it was used directly from elsewhere.

Thanks for the suggestion.

**Comment by tstuefe** [comment]
Looks still good.

..Thomas

**Review by gerard-ziemski — APPROVED** [review]
Looks good, thanks!

**Comment by dholmes-ora** [comment]
Thanks Gerard!

**Pull request closed without merging**